### DISCUSSION TITLE
Multi-GPU Tensor Initialization Question

### DISCUSSION THREAD

**Discussion by xsys-technology** [Lightning-AI/pytorch-lightning#11774] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
The documentation advises the usage of '_type_as_' when initializing new tensors in multi-gpu settings:

> [https://pytorch-lightning.readthedocs.io/en/stable/advanced/multi_gpu.html#init-tensors-using-type-as-and-register-buffer](url)
When you need to create a new tensor, use type_as. This will make your code scale to any arbitrary number of GPUs or TPUs with Lightning.

The example shows a case where a new tensor is initialized inside a LightningModule forward function:

> def forward(self, x):
         z = torch.Tensor(2, 3)
         z = z.type_as(x)

Presumably _x_ is a tensor that has already been initialized on the target gpu.

My question is what to do in the case where we want to initialize a new tensor on the target gpu, and we **do not** have access to a tensor that has already been initialized on the target gpu? 

For example, how does one properly initialize a new tensor when it is created inside a Dataset constructor that is instantiated during LightningDataModule _setup()_?

> class SomeDataModule(LightningDataModule):
...
    def setup(self, stage: Optional[str] = None):
        if stage in (None, "fit"):
            dataset = SomeDataset()
...

where:

> class SomeDataset(Dataset):
    def __init__(self):
        self.some_tensor = torch.Tensor(2,3)

Will using _type_as_ on the new tensor initialize the data on the target gpu?

> self.some_tensor = self.some_tensor.type_as(self.some_tensor)

Or is a different approach necessary? (e.g. [register_buffer()](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer))

**Comment by rohitgr7** [comment, marked as answer]
if it's part of the dataset, it's already moved to the target device when a batch is created while iterating over the dataset.

**Reply by xsys-technology** [reply]
This is a custom dataset where I'm trying to read data from a disk to populate a tensor during its constructor Dataset.__init()__

The goal is to have my data already on each gpu so that calls to __get_item()__ run as fast as possible.

It is my understanding that batches are not created until after __get_item()__ is called.

**Reply by xsys-technology** [reply]
Another thing to consider (which I haven't mentioned) is that I'm also trying to create the whole batch inside each __get_item()__ call in order to handle complex training examples (i.e. dictionaries of tensors).

>     def __getitem__(self, idx):
        return {
            'input_ids': input_ids,
            'attention_mask': attention_mask,
            'labels': labels
        }

**Reply by rohitgr7** [reply]
inside setup, you can access the current device using `self.device`, so you can pass it into the Dataset as a parameter and load data to the device.

also, with this, I think errors will come up if you use `num_workers > 1`, but not 100% sure.

**Reply by rohitgr7** [reply]
batches are created inside DataLoader's collate_fn. Dataset returns the data based on the individual indices provider by DataLoader's Sampler, which is sent to the collate_fn, where these samples are combined to create a batch.

**Reply by xsys-technology** [reply]
If things don't seem to be working (it's still going to be a couple hours before I'm ready to test) I'll give self.device a try.  This would however seem to go against the advice [preparing-your-code](https://pytorch-lightning.readthedocs.io/en/stable/advanced/multi_gpu.html#preparing-your-code).

I was hoping to let lighting automatically handle num_workers correctly, but I'll play around with that if need be.

I was also hoping that lighting will auto-magically handle collate_fn correctly. Again, if need be I can write a custom collate_fn (thank you for reminding me :)

**Reply by rohitgr7** [reply]
> This would however seem to go against the advice [preparing-your-code](https://pytorch-lightning.readthedocs.io/en/stable/advanced/multi_gpu.html#preparing-your-code).

Also, those are just suggestions for best practices, especially `.cuda`, but since here you don't have a reference to a tensor which already exists on the device, it's ok to use `.to(device)`

> I was also hoping that lighting will auto-magically handle collate_fn correctly.

no, you don't have to write a custom collate_fn. PT will handle that as long as [these](https://github.com/pytorch/pytorch/blob/8559d39cf032a9118c470ebe3a5047f89d94fe5c/torch/utils/data/_utils/collate.py#L93-L102) are returned from Dataset get_item.

> I was hoping to let lighting automatically handle num_workers correctly, 

this is up to the user to set.

**Reply by xsys-technology** [reply]
Thank you for the good info @rohitgr7 . Seems like I should be able to get things working as planned by using it. I'll post back after I have a chance to test.

**Reply by xsys-technology** [reply]
After taking a close look at the default collator code, and realizing it did exactly what I wanted, I ended up simply letting it create  batches for me by feeding it individual dictionary-of-tensor examples returned through __get_item()__ . All tensors are initialized efficiently (I didn't have to add my own .to(device)) and my gpu's are crunching away with a consistent 96% utilization with a dozen worker threads feeding them. In short, I'm happy :)
